dspace-api/src/main/java/org/dspace/app/util/DCInput.java

```diff
@@ -536,13 +559,51 @@ public List<String> getExternalSources() {
         return externalSources;
     }
 
+    /**
+     * Is user allowed for particular ACL action on this input field in given Context?
+     *
+     * @param c current Context, load the user data based on the current Context
+     * @param action read/write
+     * @return true if allowed, false otherwise
+     */
+    public boolean isAllowedAction(Context c, int action) {
+        return acl.isAllowedAction(c, action);
+    }
+
     public boolean isQualdropValue() {
         if ("qualdrop_value".equals(getInputType())) {
             return true;
         }
         return false;
     }
 
+    public ComplexDefinition getComplexDefinition() {
+        return this.complexDefinition;
+    }
+
+    /**
+     * Convert complex definition HashMap to the ordered JSON string
```

Comment:
The order is the insertion order (the order in which they are in the input xml) and this is the display order in the FE?
How is the serialization order defined?
If I have inputs firstName, lastName, email. eg. Tomas;Fuk;email@example.com, and then someone comes and wants it to be shown (in FE) as lastName, firstName, email. What do I do? Swap it in the input xml and then do a database update (so it matches Fuk;Tomas;email@example.com)?
The serialization order used to be defined by the sorting on names of the inputs; so if I named them 1_firstName,2_lastName,3_email, the complex field was serialized in that order (1;2;3) no matter how I rearranged the input-forms.xml. It is a rather nasty workaround...
I've noticed in the database that, for example, the `dspace.agreements.cookies` field is stored as a "JSON string." Is that an option? (It wasn't previously. I didn't want to parse JSON in XSLT. XSLTs are gone from the FE; still, they are there in the OAI "app"; XPath 3.1 has a parse-json function; do we have an XSLT processor that supports that?)